drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,270 @@
+    }
+
+    flash
+        .bank2()
+        .keyr
+        .write(|w| unsafe { w.keyr().bits(FLASH_KEY1) });
+    flash
+        .bank2()
+        .keyr
+        .write(|w| unsafe { w.keyr().bits(FLASH_KEY2) });
+
+    flash
+        .optkeyr()
+        .write(|w| unsafe { w.optkeyr().bits(FLASH_OPT_KEY1) });
+    flash
+        .optkeyr()
+        .write(|w| unsafe { w.optkeyr().bits(FLASH_OPT_KEY2) });
+}
+
+// See RM0433 Rev 7 section 4.3.13
+fn swap_banks() -> Result<(), UpdateError> {
+    let flash = unsafe { &*device::FLASH::ptr() };
+
+    ringbuf_entry!(Trace::FinishStart);
+    if flash.optsr_cur().read().swap_bank_opt().bit() {
+        flash
+            .optsr_prg()
+            .modify(|_, w| w.swap_bank_opt().clear_bit());
+    } else {
+        flash.optsr_prg().modify(|_, w| w.swap_bank_opt().set_bit());
+    }
+
+    flash.optcr().modify(|_, w| w.optstart().set_bit());
+
+    loop {
+        if !flash.optsr_cur().read().opt_busy().bit() {
+            break;
+        }
+    }
+
+    ringbuf_entry!(Trace::FinishEnd);
+    Ok(())
+}
+
+const FLASH_TIMEOUT: usize = 10000;
+
+fn poll_flash_done() -> Result<(), UpdateError> {
+    let flash = unsafe { &*device::FLASH::ptr() };
+
+    let mut cnt = 0;
+    let mut seen = false;
+
+    loop {
+        if flash.bank2().sr.read().qw().bit() {
+            seen = true;
+            cnt = 0;
+        }
+
+        if seen && !flash.bank2().sr.read().qw().bit() {
+            break;
+        } else {
+            cnt += 1;
+        }
+
+        if cnt == FLASH_TIMEOUT {
+            break;
+        }
+    }
+
+    if cnt == FLASH_TIMEOUT {
+        Err(UpdateError::Timeout)
+    } else {
+        Ok(())
+    }
+}
+
+fn bank_erase() -> Result<(), UpdateError> {
+    let flash = unsafe { &*device::FLASH::ptr() };
+
+    ringbuf_entry!(Trace::EraseStart);
+    flash
+        .bank2()
+        .cr
+        .write(|w| w.start().set_bit().ber().set_bit());
+
+    let b = poll_flash_done();
+    ringbuf_entry!(Trace::EraseEnd);
+    b
+}
+
+// RM0433 Rev 7 section 4.3.9
+// Following Single write sequence
+fn write_word(word_number: usize, bytes: &[u8]) -> Result<(), UpdateError> {
+    let flash = unsafe { &*device::FLASH::ptr() };
+
+    ringbuf_entry!(Trace::WriteStart);
+    flash.bank2().cr.write(|w| w.pg().set_bit());
+
+    if word_number > BANK_WORD_LIMIT {
+        return Err(UpdateError::BadBlockSize);
+    }
+
+    let start = BANK_ADDR + (word_number * FLASH_WORD_BYTES) as u32;
+
+    if bytes.len() != FLASH_WORD_BYTES {
+        return Err(UpdateError::BadBlockSize);
+    }
+
+    for i in 0..FLASH_CPU_WORDS {
+        unsafe {
+            let byte_offset = i * 4;
+            let mut word: [u8; 4] = [0; 4];
+            word.clone_from_slice(&bytes[byte_offset..byte_offset + 4]);
+
+            core::ptr::write_volatile(
+                (start + (i * 4) as u32) as *mut u32,
+                u32::from_ne_bytes(word),
+            );
+        }
+    }
+
+    poll_flash_done()?;
```

Comment:
In `bank_erase`, we logged the end state in the ringbuf regardless of `poll_flash_done` status.  Here, we return early if `poll_flash_done` returns an error, without logging to the ringbuf.

I don't feel too strongly one way or the other, but they should probably be consistent.  Maybe even log `poll_flash_done`'s result in the ringbuf, e.g. 
```rust
let b = poll_flash_done();
ringbuf_entry!(Trace::WriteEnd(b));
b
```

Reply:
This is more a side effect of the ringbuf being used for debugging (and wondering what was up with erase at one point). I'll see about making it consistent